src/vm/comsynchronizable.cpp

```diff
@@ -1537,15 +1537,58 @@ void QCALLTYPE ThreadNative::nativeInitCultureAccessors()
     END_QCALL;
 }
 
+typedef HRESULT(WINAPI *pfnSetThreadDescription)(HANDLE hThread, PCWSTR lpThreadDescription);
+extern pfnSetThreadDescription g_pfnSetThreadDescription;
+
+HRESULT WINAPI InitializeSetThreadDescription(HANDLE hThread, PCWSTR lpThreadDescription)
+{
+    pfnSetThreadDescription func = NULL;
+
+#ifndef FEATURE_PAL
+    HMODULE hKernel32 = WszLoadLibrary(W("kernel32.dll"));
+    if (hKernel32 != NULL)
+    {
+        func = (pfnSetThreadDescription)GetProcAddress(hKernel32, "SetThreadDescription");
+    }
+
+    if (func != NULL) // method is only available with Windows 10 Creators Update
+    {
+        g_pfnSetThreadDescription = func;
+        return func(hThread, lpThreadDescription);
+    }
+    else
+#endif
+    { 
+        return NOERROR;
+    }
+}
+
+pfnSetThreadDescription g_pfnSetThreadDescription = &InitializeSetThreadDescription;
 
 void QCALLTYPE ThreadNative::InformThreadNameChange(QCall::ThreadHandle thread, LPCWSTR name, INT32 len)
 {
     QCALL_CONTRACT;
 
     BEGIN_QCALL;
 
+    
     Thread* pThread = &(*thread);
 
+    // Set on Windows 10 Creators Update and later machines the unmanaged thread name as well. That will show up in ETW traces and debuggers which is very helpful
+    // if more and more thread get a meaningful name
+    if (len > 0 && name != NULL)
+    {
+        HRESULT lret = g_pfnSetThreadDescription(pThread->GetThreadHandle(), name);
+#ifdef _DEBUG
+        if (FAILED(lret))
+        {
+
+            auto last = ::GetLastError();
```

Comment:
Please note that the printf is only there for _DEBUG builds. Release builds will silently ignore that error. If printf is for debug not allowed I can remove it but I have seen plenty of printfs at other places so I thought it would be common. 
Good catch with the GetLastError. 
I will make the changes and submit an update.